crowbar_framework/app/models/crowbar_service.rb

```diff
@@ -194,11 +194,14 @@ def prepare_nodes_for_crowbar_upgrade
 
     # To all nodes, add a new role which prepares them for the upgrade
     nodes_to_upgrade = []
+    not_ready_for_upgrade = []
     all_nodes = NodeObject.all
-    not_ready = all_nodes.select { |n| !n.admin? && n.state != "ready" }.map { |n| n.name }
+    all_nodes.each do |n|
+      not_ready_for_upgrade.push n.name unless !n.admin? && !%w(ready crowbar_upgrade).include?(n.state)
```

Comment:
I think your test is wrong, the correct way would be either:
`if !n.admin? && !%w(ready crowbar_upgrade).include?(n.state)`
or
`unless n.admin? || %w(ready crowbar_upgrade).include?(n.state)`